backend/pkg/schema/client.go

```diff
@@ -1,793 +1,390 @@
-	}
+const mainProtoFilename = "__console_tmp.proto"
 
-	if cfg.TLS.Enabled {
-		tlsConfig, err := cfg.TLS.TLSConfig()
-		if err != nil {
-			return nil, fmt.Errorf("failed to create client tls config: %w", err)
-		}
+// CachedClient provides schema management with caching for efficient reuse.
+// It retrieves and parses Avro, Protobuf, and JSON schemas from a Schema Registry,
+// utilizing in-memory caches to minimize redundant fetches and compilations.
+type CachedClient struct {
+	schemaClientFactory schema.ClientFactory
+	cacheNamespace      func(context.Context) (string, error)
 
-		transport := &http.Transport{TLSClientConfig: tlsConfig}
+	standardImportsResolver func(protocompile.Resolver) protocompile.Resolver
 
-		client.SetTransport(transport)
-	}
+	schemaCache        *cache.Cache[string, sr.Schema]
+	subjectSchemaCache *cache.Cache[string, sr.SubjectSchema]
 
-	return &Client{
-		cfg:    cfg,
-		client: client,
-	}, nil
+	avroSchemaCache  *cache.Cache[string, avro.Schema]
+	protoSchemaCache *cache.Cache[string, linker.Files]
 }
 
-// SchemaResponse is the schema of the GET /schemas/ids/${id} endpoint.
-// `schema.Response` seems a little too vague for me.
-//
-//nolint:revive // This is stuttering when calling this with the pkg name, but without that the
-type SchemaResponse struct {
-	Schema     string            `json:"schema"`
-	References []SchemaReference `json:"references,omitempty"`
+// Client defines the interface for a schema client implementation.
+type Client interface {
+	AvroSchemaByID(ctx context.Context, id int) (avro.Schema, error)
+	ProtoFilesByID(ctx context.Context, id int) (linker.Files, string, error)
+	ParseAvroSchemaWithReferences(ctx context.Context, schema sr.Schema, schemaCache *avro.SchemaCache) (avro.Schema, error)
+	ParseJSONSchema(ctx context.Context, sch sr.Schema) (*jsonschema.Schema, error)
+	SchemaByID(ctx context.Context, id int) (sr.Schema, error)
+	SchemaByVersion(ctx context.Context, subject string, id int) (sr.SubjectSchema, error)
 }
 
-// GetSchemaByID returns the schema string identified by the input ID.
-// id (int) – the globally unique identifier of the schema
-func (c *Client) GetSchemaByID(ctx context.Context, id uint32) (*SchemaResponse, error) {
-	url := fmt.Sprintf("/schemas/ids/%d", id)
-	req := c.client.R().
-		SetContext(ctx).
-		SetResult(&SchemaResponse{})
+// Ensure CachedClient implements the Client interface.
+var _ Client = (*CachedClient)(nil)
 
-	res, err := req.Get(url)
-	if err != nil {
-		return nil, fmt.Errorf("get schema by id request failed: %w", err)
+// NewCachedClient initializes and returns a new CachedClient instance with the
+// provided schema client factory and cache namespace function. It sets up
+// caching with specific settings.
+func NewCachedClient(schemaClientFactory schema.ClientFactory, cacheNamespaceFn func(context.Context) (string, error)) (*CachedClient, error) {
+	cacheSettings := []cache.Opt{
+		cache.MaxAge(30 * time.Second),
```

Comment:
If I am understanding the code correctly... Now we do not have a long lived cache of all protobufs that we refresh at some interval from SR? But our registry is more dynamic based on this cache... Is this TTL _too short_?

Reply:
That's correct. I think fetching a schema is relatively cheap. I feel like 30s is a good balance because

1. For deserializing messages we won't pull hundreds of times per second
2. If someone (hard) deletes a schema we'll be up-to-date within a reasonable time. If it's needed again after 30s, we can pull again no problem

Any specific cases you have in mind where it may be problematic?

Reply:
I don't know maybe 30s is fine.